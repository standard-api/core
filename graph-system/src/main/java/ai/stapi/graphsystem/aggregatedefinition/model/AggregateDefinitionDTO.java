package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.ArrayList;
import java.util.List;

public class AggregateDefinitionDTO {

  private String id;
  private String name;
  private String description;
  private List<CommandHandlerDefinitionDTO> command;
  private StructureDefinitionId structure;

  protected AggregateDefinitionDTO() {
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      List<CommandHandlerDefinitionDTO> command,
      StructureDefinitionId structure
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.command = command;
    this.structure = structure;
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      StructureDefinitionId structure
  ) {
    this(id, name, description, new ArrayList<>(), structure);
  }

  public AggregateDefinitionDTO addCommandHandlerDefinition(
      CommandHandlerDefinitionDTO commandHandlerDefinition
  ) {
    this.command.add(commandHandlerDefinition);
    return this;
  }

  @JsonIgnore
  public boolean containsOperation(String operationId) {
    return this.command.stream().anyMatch(
        someCommand -> someCommand.getOperation().getId().equals(operationId)
    );
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  public List<CommandHandlerDefinitionDTO> getCommand() {
    return command;
  }

  public StructureDefinitionId getStructure() {
    return structure;
  }
}