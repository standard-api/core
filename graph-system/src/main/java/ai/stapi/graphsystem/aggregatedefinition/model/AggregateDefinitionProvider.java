package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import java.util.List;

public interface AggregateDefinitionProvider {

  List<AggregateDefinitionDTO> provideAll();

  AggregateDefinitionDTO provide(String aggregateType) throws CannotProvideAggregateDefinition;

  default AggregateDefinitionDTO getAggregateForOperation(
      String operationDefinitionId
  ) throws CannotProvideAggregateDefinition {
    var foundAggregates = this.provideAll().stream()
        .filter(aggregate -> aggregate.containsOperation(operationDefinitionId))
        .toList();

    if (foundAggregates.isEmpty()) {
      throw CannotProvideAggregateDefinition.becauseThereIsNoCommandHandlerForOperation(
          operationDefinitionId
      );
    }
    return foundAggregates.get(0);
  }

  default boolean containsAggregateForOperation(String operationDefinitionId) {
    return this.provideAll().stream().anyMatch(
        aggregate -> aggregate.containsOperation(operationDefinitionId)
    );
  }
}
