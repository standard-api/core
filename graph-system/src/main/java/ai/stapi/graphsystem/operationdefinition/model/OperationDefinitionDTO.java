package ai.stapi.graphsystem.operationdefinition.model;

import ai.stapi.graphsystem.operationdefinition.exceptions.CannotMergeOperationDefinition;
import ai.stapi.graphsystem.operationdefinition.exceptions.InvalidOperationDefinition;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;

public class OperationDefinitionDTO {

  private String id;
  private String name;
  private String status;
  private String kind;
  private List<String> description;
  private String code;
  private List<String> resource;
  private boolean system;
  private boolean type;
  private boolean instance;
  private List<ParameterDTO> parameter;

  private OperationDefinitionDTO() {
  }

  public OperationDefinitionDTO(
      String id,
      String name,
      String status,
      String kind,
      String description,
      String code,
      List<String> resource,
      boolean system,
      boolean type,
      boolean instance,
      List<ParameterDTO> parameter
  ) {
    this(
        id,
        name,
        status,
        kind,
        List.of(description),
        code,
        resource,
        system,
        type,
        instance,
        parameter
    );
  }

  private OperationDefinitionDTO(
      String id,
      String name,
      String status,
      String kind,
      List<String> description,
      String code,
      List<String> resource,
      boolean system,
      boolean type,
      boolean instance,
      List<ParameterDTO> parameter
  ) {
    this.id = id;
    this.name = name;
    this.status = status;
    this.kind = kind;
    this.description = description;
    this.code = code;
    this.resource = resource;
    this.system = system;
    this.type = type;
    this.instance = instance;
    this.parameter = parameter;
  }

  public static OperationDefinitionDTO bareBone(
      String id,
      List<ParameterDTO> parameter
  ) {
    return new OperationDefinitionDTO(
        id,
        null,
        null,
        null,
        "",
        null,
        List.of(),
        false,
        false,
        false,
        parameter
    );
  }

  public void addParameters(List<ParameterDTO> parameters) {
    this.parameter.addAll(parameters);
  }

  public OperationDefinitionDTO merge(OperationDefinitionDTO other) {
    if (!Objects.equals(this.getId(), other.getId())) {
      throw CannotMergeOperationDefinition.becauseSomeFieldIsNotSame(
          "id",
          this.getId(),
          other.getId()
      );
    }
    if (!Objects.equals(this.getCode(), other.getCode())) {
      throw CannotMergeOperationDefinition.becauseSomeFieldIsNotSame(
          "code",
          this.getCode(),
          other.getCode()
      );
    }
    if (!Objects.equals(this.getName(), other.getName())) {
      throw CannotMergeOperationDefinition.becauseSomeFieldIsNotSame(
          "name",
          this.getName(),
          other.getName()
      );
    }
    if (!Objects.equals(this.getStatus(), other.getStatus())) {
      throw CannotMergeOperationDefinition.becauseSomeFieldIsNotSame(
          "status",
          this.getStatus(),
          other.getStatus()
      );
    }
    if (!Objects.equals(this.getKind(), other.getKind())) {
      throw CannotMergeOperationDefinition.becauseSomeFieldIsNotSame(
          "kind",
          this.getKind(),
          other.getKind()
      );
    }
    var mergedResources = new HashSet<>(this.getResource());
    mergedResources.addAll(other.getResource());

    var mergedParameters = new ArrayList<>(this.getParameter());
    mergedParameters.addAll(other.getParameter());

    var deduplicatedParams = mergedParameters.stream()
        .collect(Collectors.groupingBy(ParameterDTO::getName))
        .values()
        .stream()
        .map(group -> group.stream().collect(Collectors.groupingBy(ParameterDTO::getType)))
        .map(group -> group.values().stream().map(
                typeGroup -> typeGroup.stream().reduce(ParameterDTO::merge).get()
            ).toList()
        ).flatMap(group -> {
          if (group.size() == 1) {
            return group.stream();
          }
          return group.stream().map(parameterDTO -> new ParameterDTO(
              String.format(
                  "%s%s",
                  parameterDTO.getName(),
                  StringUtils.capitalize(parameterDTO.getType())
              ),
              parameterDTO.getUse(),
              parameterDTO.getMin(),
              parameterDTO.getMax(),
              parameterDTO.getType(),
              parameterDTO.getReferencedFrom(),
              parameterDTO.getTargetProfileReference()
          ));
        }).toList();

    var mergedDescriptions = new ArrayList<>(this.description);
    mergedDescriptions.addAll(other.description);
    return new OperationDefinitionDTO(
        this.getId(),
        this.getName(),
        this.getStatus(),
        this.getKind(),
        mergedDescriptions,
        this.getCode(),
        mergedResources.stream().toList(),
        this.isSystem() || other.isSystem(),
        this.isType() || other.isType(),
        this.isInstance() || other.isInstance(),
        deduplicatedParams
    );
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getStatus() {
    return status;
  }

  public String getKind() {
    return kind;
  }

  public String getDescription() {
    if (this.description.isEmpty()) {
      return "";
    }
    if (this.description.size() == 1) {
      return this.description.get(0);
    }
    return String.format(
        "This automatic adding Operation was merged, because two ComplexType unionType members "
            + "had fieldName with same name, but different type. Original descriptions: %n%s",
        String.join("\n", this.description)
    );
  }

  private void setDescription(String description) {
    this.description = List.of(description);
  }

  public String getCode() {
    return code;
  }

  public List<String> getResource() {
    return resource;
  }

  public boolean isSystem() {
    return system;
  }

  public boolean isType() {
    return type;
  }

  public boolean isInstance() {
    return instance;
  }

  public List<ParameterDTO> getParameter() {
    return parameter;
  }

  @JsonIgnore
  public ParameterDTO getParameter(String parameterName) {
    return this.getParameter().stream()
        .filter(paramater -> paramater.getName().equals(parameterName))
        .findAny()
        .orElseThrow(
            () -> new InvalidOperationDefinition(
                "Parameter '%s' not found.".formatted(parameterName)
            )
        );
  }

  @Override
  public String toString() {
    return "OperationDefinitionDTO{" +
        "id='" + id + '\'' +
        ", name='" + name + '\'' +
        ", status='" + status + '\'' +
        ", kind='" + kind + '\'' +
        ", description='" + description + '\'' +
        ", code='" + code + '\'' +
        ", resource=" + resource +
        ", system=" + system +
        ", type=" + type +
        ", instance=" + instance +
        ", parameter=" + parameter +
        '}';
  }

  public static class ParameterDTO {

    private String name;
    private String use;
    private Integer min;
    private String max;
    private String type;
    private List<ReferencedFrom> referencedFrom;
    private List<StructureDefinitionId> targetProfileReference;

    private ParameterDTO() {
    }

    private ParameterDTO(
        String name,
        String use,
        Integer min,
        String max,
        String type,
        List<ReferencedFrom> referencedFrom,
        List<StructureDefinitionId> targetProfileReference
    ) {
      this.name = name;
      this.use = use;
      this.min = min;
      this.max = max;
      this.type = type;
      this.referencedFrom = referencedFrom;
      this.targetProfileReference = targetProfileReference;
    }

    public ParameterDTO(
        String name,
        String use,
        Integer min,
        String max,
        String type,
        ReferencedFrom referencedFrom,
        List<StructureDefinitionId> targetProfileReference
    ) {
      this.name = name;
      this.use = use;
      this.min = min;
      this.max = max;
      this.type = type;
      this.referencedFrom = new ArrayList<>(List.of(referencedFrom));
      this.targetProfileReference = targetProfileReference;
    }

    public ParameterDTO merge(ParameterDTO other) {
      if (!Objects.equals(this.getName(), other.getName())) {
        throw CannotMergeOperationDefinition.becauseSomeParameterHasSomeFieldWhichIsNotSame(
            this.getName(),
            "name",
            this.getName(),
            other.getName()
        );
      }
      if (!Objects.equals(this.getUse(), other.getUse())) {
        throw CannotMergeOperationDefinition.becauseSomeParameterHasSomeFieldWhichIsNotSame(
            this.getName(),
            "use",
            this.getUse(),
            other.getUse()
        );
      }
      if (!Objects.equals(this.getMin(), other.getMin())) {
        throw CannotMergeOperationDefinition.becauseSomeParameterHasSomeFieldWhichIsNotSame(
            this.getName(),
            "min",
            this.getMin(),
            other.getMin()
        );
      }
      if (!Objects.equals(this.getMax(), other.getMax())) {
        throw CannotMergeOperationDefinition.becauseSomeParameterHasSomeFieldWhichIsNotSame(
            this.getName(),
            "max",
            this.getMax(),
            other.getMax()
        );
      }
      if (!Objects.equals(this.getType(), other.getType())) {
        throw CannotMergeOperationDefinition.becauseSomeParameterHasSomeFieldWhichIsNotSame(
            this.getName(),
            "type",
            this.getType(),
            other.getType()
        );
      }
      var mergedReferencedFrom = new ArrayList<>(this.getReferencedFrom());
      mergedReferencedFrom.addAll(other.getReferencedFrom());

      var mergedTargetProfileReferences = new ArrayList<>(this.getTargetProfileReference());
      mergedTargetProfileReferences.addAll(other.getTargetProfileReference());

      return new ParameterDTO(
          this.getName(),
          this.getUse(),
          this.getMin(),
          this.getMax(),
          this.getType(),
          mergedReferencedFrom.stream()
              .collect(Collectors.groupingBy(ReferencedFrom::getSource))
              .values()
              .stream()
              .map(group -> group.get(0))
              .toList(),
          mergedTargetProfileReferences.stream()
              .collect(Collectors.groupingBy(StructureDefinitionId::getId))
              .values()
              .stream()
              .map(group -> group.get(0))
              .toList()
      );
    }

    public String getName() {
      return name;
    }

    public String getUse() {
      return use;
    }

    public Integer getMin() {
      return min;
    }

    public String getMax() {
      return max;
    }

    public String getType() {
      return type;
    }

    public List<ReferencedFrom> getReferencedFrom() {
      return referencedFrom;
    }

    @JsonIgnore
    public ReferencedFrom getSingleReferencedFrom() {
      if (this.referencedFrom.size() != 1) {
        throw new InvalidOperationDefinition(
            String.format(
                "Every parameter should always have exactly one referencedFrom value.%n%s",
                String.format("Operation name: '%s'", this.name)
            )
        );
      }
      return referencedFrom.get(0);
    }

    public List<StructureDefinitionId> getTargetProfileReference() {
      return targetProfileReference;
    }

    @JsonIgnore
    public boolean isList() {
      try {
        var number = Integer.parseInt(this.max);
        return number > 1;
      } catch (RuntimeException ignore) {
      }
      return this.max.equals("*");
    }

    @Override
    public String toString() {
      return "Parameter{" +
          "name='" + name + '\'' +
          ", use='" + use + '\'' +
          ", min=" + min +
          ", max='" + max + '\'' +
          ", type=" + type +
          '}';
    }

    public static class ReferencedFrom {

      // Location in resource (into what the parameter gets mapped to) eg.: Patient.name
      private String source;

      private ReferencedFrom() {
      }

      public ReferencedFrom(String source) {
        this.source = source;
      }

      public String getSource() {
        return source;
      }
    }
  }
}
