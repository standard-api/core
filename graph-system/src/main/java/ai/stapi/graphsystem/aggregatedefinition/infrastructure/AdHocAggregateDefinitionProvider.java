package ai.stapi.graphsystem.aggregatedefinition.infrastructure;

import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionProvider;
import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import ai.stapi.schema.adHocLoaders.GenericAdHocModelDefinitionsLoader;
import ai.stapi.schema.scopeProvider.ScopeCacher;
import ai.stapi.schema.scopeProvider.ScopeOptions;
import ai.stapi.graphsystem.operationdefinition.exceptions.CannotProvideOperationDefinition;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

public class AdHocAggregateDefinitionProvider implements AggregateDefinitionProvider {

  private final GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader;
  private final ScopeCacher scopeCacher;
  private final OperationDefinitionProvider operationDefinitionProvider;

  public AdHocAggregateDefinitionProvider(
      GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader,
      ScopeCacher scopeCacher,
      OperationDefinitionProvider operationDefinitionProvider
  ) {
    this.genericAdHocModelDefinitionsLoader = genericAdHocModelDefinitionsLoader;
    this.scopeCacher = scopeCacher;
    this.operationDefinitionProvider = operationDefinitionProvider;
  }

  @Override
  public List<AggregateDefinitionDTO> provideAll() {
    return this.getAggregateDefinitions();
  }

  @Override
  public AggregateDefinitionDTO provide(
      String aggregateType
  ) throws CannotProvideOperationDefinition {
    return this.getAggregateDefinitions().stream()
        .filter(aggregate -> aggregate.getName().equals(aggregateType))
        .findFirst()
        .orElseThrow(() -> CannotProvideAggregateDefinition.becauseItDoesNotExist(aggregateType));
  }

  private List<AggregateDefinitionDTO> getAggregateDefinitions() {
    return this.scopeCacher.getCachedOrCompute(
        AdHocAggregateDefinitionProvider.class,
        this::load
    );
  }

  private List<AggregateDefinitionDTO> load(ScopeOptions scopeOptions) {
    var createAggregates = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "CreateAggregateDefinition",
            AggregateDefinitionDTO.class
        ).stream()
        .toList();

    var additionalCommandHandlers = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "AddCommandOnAggregateDefinition",
            AddCommandOnAggregateDefinitionDTO.class
        ).stream()
        .toList();

    this.ensureAdditionalCommandAreValid(scopeOptions, createAggregates, additionalCommandHandlers);

    var groupedCommands = additionalCommandHandlers
        .stream()
        .collect(Collectors.groupingBy(AddCommandOnAggregateDefinitionDTO::getId));

    var aggregatesWithAddedCommands = createAggregates.stream().map(
        aggregate -> {
          var addedCommands = groupedCommands.get(aggregate.getId());
          if (addedCommands == null) {
            return aggregate;
          }
          addedCommands.stream()
              .map(AddCommandOnAggregateDefinitionDTO::getCommand)
              .flatMap(List::stream)
              .forEach(aggregate::addCommandHandlerDefinition);

          return aggregate;
        }
    ).toList();

    var additionalModifications = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "AddModificationOnAggregateDefinitionCommandEventFactory",
            AddModificationOnAggregateDefinitionDTO.class
        ).stream()
        .toList();

    this.ensureAdditionalModificationsAreValid(
        scopeOptions,
        aggregatesWithAddedCommands,
        additionalModifications
    );

    var groupedModifications = additionalModifications
        .stream()
        .collect(Collectors.groupingBy(AddModificationOnAggregateDefinitionDTO::getId));

    var aggregatesWithAddedModifications = createAggregates.stream().map(
        aggregate -> {
          var addedModifications = groupedModifications.get(aggregate.getId());
          if (addedModifications == null) {
            return aggregate;
          }
          addedModifications.forEach(addModification -> {
            var foundEventFactory = aggregate.getCommand()
                .stream()
                .flatMap(command -> command.getEventFactory().stream())
                .filter(eventFactory -> eventFactory.getId().equals(addModification.getEventFactoryId()))
                .findFirst()
                .get();

            foundEventFactory.addModifications(addModification.getModification());
          });

          return aggregate;
        }
    ).toList();

    return aggregatesWithAddedModifications.stream().map(aggregate -> new AggregateDefinitionDTO(
        aggregate.getId(),
        aggregate.getName(),
        aggregate.getDescription(),
        aggregate.getCommand().stream().map(command -> new CommandHandlerDefinitionDTO(
            this.operationDefinitionProvider.provide(command.getOperation().getId()),
            command.getCreationalPolicy(),
            command.getEventFactory()
        )).toList(),
        aggregate.getStructure()
    )).toList();
  }

  private void ensureAdditionalCommandAreValid(
      ScopeOptions scopeOptions,
      List<AggregateDefinitionDTO> createAggregates,
      List<AddCommandOnAggregateDefinitionDTO> additionalCommandHandlers
  ) {
    var addCommandWithoutTarget = additionalCommandHandlers.stream()
        .filter(map -> this.isNotContainedInFoundAggregates(createAggregates, map))
        .toList();

    if (!addCommandWithoutTarget.isEmpty()) {
      throw new CannotProvideAggregateDefinition(
          String.format(
              "Cannot provide aggregate definitions for scope: %s with tags: %s.%n%s%n%s",
              scopeOptions.getScopes(),
              scopeOptions.getTags(),
              "There were some AddCommandOnAggregateDefinition JSONs which eiter did "
                  + "not have id specified or it was missing in found AggregateDefinitions.",
              "Missing ids: " + addCommandWithoutTarget.stream()
                  .map(AddCommandOnAggregateDefinitionDTO::getId)
                  .toList()
          )
      );
    }
  }

  private boolean isNotContainedInFoundAggregates(
      List<AggregateDefinitionDTO> createAggregates,
      AddCommandOnAggregateDefinitionDTO addCommand
  ) {
    return addCommand.getId() == null || createAggregates.stream().noneMatch(
        aggregate -> aggregate.getId().equals(addCommand.getId())
    );
  }

  private void ensureAdditionalModificationsAreValid(
      ScopeOptions scopeOptions,
      List<AggregateDefinitionDTO> aggregates,
      List<AddModificationOnAggregateDefinitionDTO> additionalModifications
  ) {
    var addModificationWithoutTarget = additionalModifications.stream()
        .filter(map -> !this.isModificationContainedInFoundAggregates(aggregates, map))
        .toList();

    if (!addModificationWithoutTarget.isEmpty()) {
      throw new CannotProvideAggregateDefinition(
          String.format(
              "Cannot provide aggregate definitions for scope: %s with tags: %s.%n%s%n%s",
              scopeOptions.getScopes(),
              scopeOptions.getTags(),
              "There were some AddModificationOnAggregateDefinition JSONs which eiter did "
                  + "not have id specified or it was missing in found AggregateDefinitions.",
              "Missing ids: " + addModificationWithoutTarget.stream()
                  .map(AddModificationOnAggregateDefinitionDTO::getId)
                  .toList()
          )
      );
    }
  }

  private boolean isModificationContainedInFoundAggregates(
      List<AggregateDefinitionDTO> aggregates,
      AddModificationOnAggregateDefinitionDTO addModification
  ) {
    if (addModification.getId() == null) {
      return false;
    }
    if (addModification.getEventFactoryId() == null) {
      return false;
    }
    var foundAggregate = aggregates.stream()
        .filter(aggregate -> aggregate.getId().equals(addModification.getId()))
        .findFirst();

    if (foundAggregate.isEmpty()) {
      return false;
    }
    return foundAggregate.get().getCommand().stream().anyMatch(
        command -> command.getEventFactory().stream().anyMatch(
            eventFactory -> eventFactory.getId().equals(addModification.getEventFactoryId())
        )
    );
  }

  private static class AddCommandOnAggregateDefinitionDTO {

    private String id;
    private List<CommandHandlerDefinitionDTO> command;

    private AddCommandOnAggregateDefinitionDTO() {
    }

    public AddCommandOnAggregateDefinitionDTO(
        String id,
        List<CommandHandlerDefinitionDTO> command
    ) {
      this.id = id;
      this.command = command;
    }

    public String getId() {
      return id;
    }

    public List<CommandHandlerDefinitionDTO> getCommand() {
      return command;
    }
  }

  private static class AddModificationOnAggregateDefinitionDTO {

    private String id;
    private String eventFactoryId;
    private List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> modification;

    private AddModificationOnAggregateDefinitionDTO() {
    }

    public AddModificationOnAggregateDefinitionDTO(
        String id,
        String eventFactoryId,
        List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> modification
    ) {
      this.id = id;
      this.eventFactoryId = eventFactoryId;
      this.modification = modification;
    }

    public String getId() {
      return id;
    }

    public String getEventFactoryId() {
      return eventFactoryId;
    }

    public List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> getModification() {
      return modification;
    }
  }
}
