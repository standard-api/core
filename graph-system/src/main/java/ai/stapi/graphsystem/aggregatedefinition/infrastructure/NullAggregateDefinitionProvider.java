package ai.stapi.graphsystem.aggregatedefinition.infrastructure;

import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionProvider;
import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import java.util.List;
import org.springframework.context.annotation.Primary;

@Primary
public class NullAggregateDefinitionProvider implements AggregateDefinitionProvider {

  @Override
  public List<AggregateDefinitionDTO> provideAll() {
    return List.of();
  }

  @Override
  public AggregateDefinitionDTO provide(String aggregateType)
      throws CannotProvideAggregateDefinition {
    throw CannotProvideAggregateDefinition.becauseItDoesNotExist(aggregateType);
  }
}
