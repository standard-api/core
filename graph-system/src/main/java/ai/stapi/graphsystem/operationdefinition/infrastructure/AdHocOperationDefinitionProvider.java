package ai.stapi.graphsystem.operationdefinition.infrastructure;

import ai.stapi.schema.adHocLoaders.GenericAdHocModelDefinitionsLoader;
import ai.stapi.schema.scopeProvider.ScopeCacher;
import ai.stapi.schema.scopeProvider.ScopeOptions;
import ai.stapi.graphsystem.operationdefinition.exceptions.CannotProvideOperationDefinition;
import ai.stapi.graphsystem.operationdefinition.exceptions.CannotProvideOperationDefinitions;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO.ParameterDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

public class AdHocOperationDefinitionProvider implements OperationDefinitionProvider {

  private final GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader;
  private final ScopeCacher scopeCacher;

  public AdHocOperationDefinitionProvider(
      GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader,
      ScopeCacher scopeCacher
  ) {
    this.genericAdHocModelDefinitionsLoader = genericAdHocModelDefinitionsLoader;
    this.scopeCacher = scopeCacher;
  }

  @Override
  public List<OperationDefinitionDTO> provideAll() {
    return this.getOperationDefinitions();
  }

  @Override
  public OperationDefinitionDTO provide(
      String operationId
  ) throws CannotProvideOperationDefinition {
    return this.getOperationDefinitions().stream()
        .filter(operation -> operation.getId().equals(operationId))
        .findFirst()
        .orElseThrow(() -> new CannotProvideOperationDefinition(operationId));
  }

  @Override
  public boolean contains(String operationId) {
    return this.getOperationDefinitions().stream().anyMatch(
        operation -> operation.getId().equals(operationId)
    );
  }

  private List<OperationDefinitionDTO> getOperationDefinitions() {
    return this.scopeCacher.getCachedOrCompute(
        AdHocOperationDefinitionProvider.class,
        this::load
    );
  }

  private List<OperationDefinitionDTO> load(ScopeOptions scopeOptions) {
    var createOperation = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "CreateOperationDefinition",
            OperationDefinitionDTO.class
        ).stream()
        .toList();

    var addParameterOnOperation = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "AddParameterOnOperationDefinition",
            AddParameterOnOperationDefinitionDTO.class
        ).stream()
        .toList();

    var addParameterWithoutTarget = addParameterOnOperation.stream()
        .filter(map -> this.isContainedInFoundOperations(createOperation, map))
        .toList();

    if (!addParameterWithoutTarget.isEmpty()) {
      throw new CannotProvideOperationDefinitions(
          String.format(
              "Cannot provide operation definitions for scope: %s with tags: %s.%n%s%n%s",
              scopeOptions.getScopes(),
              scopeOptions.getTags(),
              "There were some AddParameterOnOperationDefinition JSONs which eiter did "
                  + "not have id specified or it was missing in found OperationDefinitions.",
              "Missing ids: " + addParameterWithoutTarget.stream()
                  .map(AddParameterOnOperationDefinitionDTO::getId)
                  .toList()
          )
      );
    }

    var groupedParams = addParameterOnOperation
        .stream()
        .collect(Collectors.groupingBy(AddParameterOnOperationDefinitionDTO::getId));

    return createOperation.stream().map(operation -> {
      var addedParameters = groupedParams.get(operation.getId());
      if (addedParameters == null) {
        return operation;
      }
      var newParameters = addedParameters.stream()
          .map(AddParameterOnOperationDefinitionDTO::getParameter)
          .flatMap(List::stream)
          .toList();

      operation.addParameters(newParameters);
      return operation;
    }).toList();
  }

  private boolean isContainedInFoundOperations(
      List<OperationDefinitionDTO> createOperation,
      AddParameterOnOperationDefinitionDTO addParameterDTO
  ) {
    return addParameterDTO.getId() == null || createOperation.stream().noneMatch(
        operation -> operation.getId().equals(addParameterDTO.getId())
    );
  }

  private static class AddParameterOnOperationDefinitionDTO {

    private String id;
    private List<ParameterDTO> parameter;

    private AddParameterOnOperationDefinitionDTO() {
    }

    public AddParameterOnOperationDefinitionDTO(String id, List<ParameterDTO> parameter) {
      this.id = id;
      this.parameter = parameter;
    }

    public String getId() {
      return id;
    }

    public List<ParameterDTO> getParameter() {
      return parameter;
    }
  }
}
