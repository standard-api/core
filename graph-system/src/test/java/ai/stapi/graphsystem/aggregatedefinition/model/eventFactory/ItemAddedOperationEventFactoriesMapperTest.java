package ai.stapi.graphsystem.aggregatedefinition.model.eventFactory;

import ai.stapi.test.SystemSchemaIntegrationTestCase;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import ai.stapi.graphsystem.operationdefinition.infrastructure.AdHocOperationDefinitionProvider;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class ItemAddedOperationEventFactoriesMapperTest extends SystemSchemaIntegrationTestCase {

  @Autowired
  private ItemAddedOperationEventFactoriesMapper mapper;

  @Autowired
  private AdHocOperationDefinitionProvider provider;

  @Autowired
  private ObjectMapper objectMapper;

  @Test
  void itCanMapFirstLevelListFieldAddOperation() {
    var operationName = "AddJurisdictionOnOperationDefinition";
    this.testMapper(operationName);
  }

  @Test
  void itCanMapListFieldOnLeafComplexFieldAddOperation() {
    var operationName = "AddElementOnStructureDefinitionDifferential";
    this.testMapper(operationName);
  }

  @Test
  void itCanMapListFieldOnListComplexFieldAddOperation() {
    var operationName = "AddCodingOnOperationDefinitionJurisdiction";
    this.testMapper(operationName);
  }

  @Test
  void itCanMapListUnionFieldAddOperation() throws JsonProcessingException {
    var inputJson = """
        {
          "id" : "AddExampleUnionFieldOnExampleResourceType",
          "code" : "AddExampleUnionFieldOnExampleResourceType",
          "description" : "Generated command for adding exampleUnionField(decimal | Extension | Reference) on ExampleResourceType Aggregate",
          "instance" : true,
          "kind" : "operation",
          "name" : "Add ExampleUnionField on ExampleResourceType",
          "status" : "draft",
          "system" : false,
          "type" : false,
          "parameter" : [ {
            "max" : "1",
            "min" : 1,
            "name" : "exampleUnionFieldDecimal",
            "type" : "decimal",
            "use" : "in",
            "referencedFrom" : [ {
              "source" : "ExampleResourceType.exampleUnionField"
            } ],
            "targetProfileReference" : [ ]
          }, {
            "max" : "1",
            "min" : 1,
            "name" : "exampleUnionFieldExtension",
            "type" : "Extension",
            "use" : "in",
            "referencedFrom" : [ {
              "source" : "ExampleResourceType.exampleUnionField"
            } ],
            "targetProfileReference" : [ ]
          }, {
            "max" : "1",
            "min" : 1,
            "name" : "exampleUnionFieldReference",
            "type" : "Reference",
            "use" : "in",
            "referencedFrom" : [ {
              "source" : "ExampleResourceType.exampleUnionField"
            } ],
            "targetProfileReference" : [ {
              "id" : "OperationDefinition"
            }, {
              "id" : "StructureDefinition"
            } ]
          } ],
          "resource" : [ "ExampleResourceType" ]
        }
        """;

    var input = this.objectMapper.readValue(inputJson, OperationDefinitionDTO.class);
    var actual = this.mapper.map(input);
    this.thenObjectApproved(actual);
  }

  private void testMapper(String operationName) {
    var input = this.provider.provide(operationName);
    var actual = this.mapper.map(input);
    this.thenObjectApproved(actual);
  }
}