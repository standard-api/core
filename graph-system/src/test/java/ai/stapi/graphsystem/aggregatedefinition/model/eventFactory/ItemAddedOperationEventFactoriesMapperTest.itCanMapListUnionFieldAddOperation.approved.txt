[ {
  "id" : "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx",
  "event" : {
    "id" : "ExampleUnionFieldOnExampleResourceTypeAdded",
    "description" : "Generated Event for adding on ExampleResourceType.",
    "name" : "ExampleUnionField on ExampleResourceType Added",
    "type" : {
      "id" : "ExampleResourceType"
    }
  },
  "modification" : [ {
    "inputValueParameterName" : "exampleUnionField",
    "kind" : "add",
    "modificationPath" : "ExampleResourceType.exampleUnionField",
    "destinationIndexParameterName" : null
  } ]
} ]